src/latexify/codegen/function_codegen.py

```diff
@@ -637,181 +128,68 @@ def visit_If(self, node: ast.If) -> str:
-
-        current_expr: ast.expr = node
-
-        while isinstance(current_expr, ast.IfExp):
-            cond_latex = self.visit(current_expr.test)
-            true_latex = self.visit(current_expr.body)
-            latex += true_latex + r", & \mathrm{if} \ " + cond_latex + r" \\ "
-            current_expr = current_expr.orelse
-
-        latex += self.visit(current_expr)
-        return latex + r", & \mathrm{otherwise} \end{array} \right."
-
-    def _reduce_stop_parameter(self, node: ast.expr) -> ast.expr:
-        """Adjusts the stop expression of the range.
-
-        This function tries to convert the syntax as follows:
-            * n + 1 --> n
-            * n + 2 --> n + 1
-            * n - (-1) --> n
-            * n - 1 --> n - 2
-
-        Args:
-            node: The target expression.
-
-        Returns:
-            Converted expression.
-        """
-        if not (
-            isinstance(node, ast.BinOp) and isinstance(node.op, (ast.Add, ast.Sub))
-        ):
-            return ast.BinOp(left=node, op=ast.Sub(), right=ast_utils.make_constant(1))
-
-        # Treatment for Python 3.7.
-        rhs = (
-            ast.Constant(value=node.right.n)
-            if sys.version_info.minor < 8 and isinstance(node.right, ast.Num)
-            else node.right
-        )
-
-        if not isinstance(rhs, ast.Constant):
-            return ast.BinOp(left=node, op=ast.Sub(), right=ast_utils.make_constant(1))
-
-        shift = 1 if isinstance(node.op, ast.Add) else -1
-
-        return (
-            node.left
-            if rhs.value == shift
-            else ast.BinOp(
-                left=node.left, op=node.op, right=ast.Constant(value=rhs.value - shift)
-            )
-        )
-
-    def _get_sum_prod_range(self, node: ast.comprehension) -> tuple[str, str] | None:
-        """Helper to process range(...) for sum and prod functions.
-
-        Args:
-            node: comprehension node to be analyzed.
-
-        Returns:
-            Tuple of following strings:
-                - lower_rhs
-                - upper
-            which are used in _get_sum_prod_info, or None if the analysis failed.
-        """
-        if not (
-            isinstance(node.iter, ast.Call)
-            and isinstance(node.iter.func, ast.Name)
-            and node.iter.func.id == "range"
-        ):
-            return None
-
-        try:
-            range_info = analyzers.analyze_range(node.iter)
-        except exceptions.LatexifyError:
-            return None
-
-        if (
-            # Only accepts ascending order with step size 1.
-            range_info.step_int != 1
-            or (
-                range_info.start_int is not None
-                and range_info.stop_int is not None
-                and range_info.start_int >= range_info.stop_int
-            )
-        ):
-            return None
+    def visit_Attribute(self, node: ast.Attribute) -> str:
+        return self._expression_codegen.visit_Attribute(node)
```

Comment:
I also think it is better not to define these visitors in `stmt` codegens. If these visitors exist, the codegen can generate LaTeX for `expr` accidentally, which is clearly not an expected behavior.

Reply:
> edit: that's the approach I've gone with for now

So is this resolved?

Reply:
I checked code and found:

```python
    def generic_visit(self, node: ast.AST) -> str:
        if isinstance(node, ast.expr):
            return self._expression_codegen.visit(node)
```

This is still involving the same problem in the second comment. I meant FunctionCodegen shouldn't implement any public functionality to process `expr` (because it is `stmt` codegen), and delegate it completely to the internal `ExpressionCodegen`, i.e., directly call `expr_codegen.visit()` in `visit_{stmt}`.

In the AST, any subtree of `expr` should be of a static type `expr` as defined in the [official doc](https://docs.python.org/3/library/ast.html), and we don't need any dynamic type checking.

Reply:
Okay I think I get it now. That's really cool actually thank you for explaining.